source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -512,8 +528,12 @@ Network::FilterStatus Filter::establishUpstreamConnection() {
     // cluster->trafficStats()->upstream_cx_none_healthy in the latter case.
     getStreamInfo().setResponseFlag(StreamInfo::CoreResponseFlag::NoHealthyUpstream);
     onInitFailure(UpstreamFailureReason::NoHealthyUpstream);
+    return Network::FilterStatus::StopIteration;
```

Comment:
This case is when there is error during Cluster creation/connection, so we sure shot want to StopIteration here rather than depend on receive_before_connect state.